test/acceptance/authz/tenants_test.go

```diff
@@ -68,93 +71,220 @@ func TestAuthZTenants(t *testing.T) {
+		helper.AssignRoleToUser(t, adminKey, *createSpecificTenantsRole.Name, customUser)
+		defer helper.RevokeRoleFromUser(t, adminKey, *createSpecificTenantsRole.Name, customUser)
+		err := createTenant(t, className, []*models.Tenant{{Name: tenant2}}, customKey)
+		require.NotNil(t, err)
+		var forbidden *clschema.TenantsCreateForbidden
+		require.True(t, errors.As(err, &forbidden))
+	})
+
+	setupRUDTests := func(role string) func() {
 		require.NoError(t, createTenant(t, className, tenants, adminKey))
-		defer deleteTenant(t, className, []string{tenants[0].Name}, adminKey)
+		helper.AssignRoleToUser(t, adminKey, role, customUser)
+		return func() {
+			defer deleteTenant(t, className, tenantNames, adminKey)
+			defer helper.RevokeRoleFromUser(t, adminKey, role, customUser)
+		}
+	}
 
-		helper.AssignRoleToUser(t, adminKey, *readTenantRole.Name, customUser)
-		defer helper.RevokeRoleFromUser(t, adminKey, *readTenantRole.Name, customUser)
+	t.Run("Tenant read and exist", func(t *testing.T) {
+		cleanup := setupRUDTests(readAllTenantsRoleName)
+		defer cleanup()
 
 		require.NoError(t, readTenant(t, className, tenants[0].Name, customKey))
-		require.NoError(t, readTenants(t, className, customKey))
 		require.NoError(t, existsTenant(t, className, tenants[0].Name, customKey))
 	})
 
-	t.Run("delete tenant", func(t *testing.T) {
-		require.NoError(t, createTenant(t, className, tenants, adminKey))
+	t.Run("Read a specific tenant with needed permission", func(t *testing.T) {
+		cleanup := setupRUDTests(readSpecificTenantRoleName)
+		defer cleanup()
 
-		helper.AssignRoleToUser(t, adminKey, *deleteTenantRole.Name, customUser)
-		defer helper.RevokeRoleFromUser(t, adminKey, *deleteTenantRole.Name, customUser)
+		require.NoError(t, readTenant(t, className, tenants[0].Name, customKey))
+		require.NoError(t, existsTenant(t, className, tenants[0].Name, customKey))
+	})
+
+	t.Run("Fail to read a specific tenant and all tenants without needed permission", func(t *testing.T) {
+		cleanup := setupRUDTests(readSpecificTenantRoleName)
+		defer cleanup()
+
+		err := readTenant(t, className, tenants[1].Name, customKey)
+		require.NotNil(t, err)
+		var forbiddenGetOne *clschema.TenantsGetOneForbidden
+		require.True(t, errors.As(err, &forbiddenGetOne))
+
+		err = existsTenant(t, className, tenants[1].Name, customKey)
+		var forbiddenExists *clschema.TenantExistsForbidden
+		require.True(t, errors.As(err, &forbiddenExists))
+
+		// tests for tenant filtering
+		res, err := readTenants(t, className, customKey)
+		require.Nil(t, err)
+		require.Len(t, res.Payload, 1)
+		require.Equal(t, tenants[0].Name, res.Payload[0].Name)
+	})
+
+	t.Run("Get specific tenant using grpc", func(t *testing.T) {
+		cleanup := setupRUDTests(readSpecificTenantRoleName)
+		defer cleanup()
+
+		_, err := readTenantGRPC(t, ctx, className, tenants[0].Name, customKey)
+		require.Nil(t, err)
+	})
+
+	t.Run("Return no tenants via grpc when one is forbidden", func(t *testing.T) {
```

Comment:
Not sure if I understand this test:
- you try to get one tenant but don't have permission
- shouldn'T we return an error instead of a list?

My understanding for getting is:
- filter for getAll => only those with permissions, if no permission the list is empty
- yes/no for getOne => if permission the tenant is returned and an error otherwise

Reply:
Yes we should but `TenantsGet` isn't `TenantsGetOne` so there is no way of fairly meeting both requirements. The issue is that we currently use `TenantsGet` to be `TenantsGetOne` when `names=[name]`, which is no longer semantically correct given the different filtering requirements between the two operations

I think what this motivates is either the introduction of that gRPC method or changing the clients to use `GET v1/schema/{className}/tenants/{tenant}` for `tenants.get_by_name` to avoid this issue. wdyt?

Reply:
ah that makes sense

> I think what this motivates is either the introduction of that gRPC method or changing the clients to use GET v1/schema/{className}/tenants/{tenant} for tenants.get_by_name to avoid this issue. wdyt?

I think the second one is good. We introduced the GRPC call because fetching 100s+ tenants over REST didn't work that well. A single one should be fine